subsys/zbus/zbus_runtime_observers.c

```diff
@@ -8,17 +8,80 @@
 
 LOG_MODULE_DECLARE(zbus, CONFIG_ZBUS_LOG_LEVEL);
 
-int zbus_chan_add_obs(const struct zbus_channel *chan, const struct zbus_observer *obs,
-		      struct zbus_observer_node *node, k_timeout_t timeout)
+#if defined(CONFIG_ZBUS_RUNTIME_OBSERVERS_NODE_ALLOC_DYNAMIC)
+
+static inline int _zbus_runtime_observer_node_alloc(struct zbus_observer_node **node,
+						    k_timeout_t timeout)
+{
+	ARG_UNUSED(timeout);
+
+	*node = k_malloc(sizeof(struct zbus_observer_node));
+
+	_ZBUS_ASSERT(*node != NULL, "could not allocate observer node the heap is full!");
+
+	if (*node == NULL) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static inline void _zbus_runtime_observer_node_free(struct zbus_observer_node *node)
+{
+	k_free(node);
+}
+
+#elif defined(CONFIG_ZBUS_RUNTIME_OBSERVERS_NODE_ALLOC_STATIC)
+
+K_MEM_SLAB_DEFINE_STATIC(_zbus_runtime_observers_slab, sizeof(struct zbus_observer_node),
+			 CONFIG_ZBUS_RUNTIME_OBSERVERS_NODE_POOL_SIZE, 8);
+
+static inline int _zbus_runtime_observer_node_alloc(struct zbus_observer_node **node,
+						    k_timeout_t timeout)
+{
+	int err = k_mem_slab_alloc(&_zbus_runtime_observers_slab, (void **)node, timeout);
+
+	_ZBUS_ASSERT(*node != NULL, "not enough runtime observer nodes in the pool. Increase the "
+				    "ZBUS_RUNTIME_OBSERVERS_NODE_POOL_SIZE");
+
+	return err;
+}
+
+static inline void _zbus_runtime_observer_node_free(struct zbus_observer_node *node)
+{
+	k_mem_slab_free(&_zbus_runtime_observers_slab, (void *)node);
+}
+
+#else
+
+static inline int _zbus_runtime_observer_node_alloc(struct zbus_observer_node **node,
+						    k_timeout_t timeout)
+{
+	ARG_UNUSED(node);
+	ARG_UNUSED(timeout);
+
+	return -ENOTSUP;
+}
+
+static inline void _zbus_runtime_observer_node_free(struct zbus_observer_node *node)
+{
+	ARG_UNUSED(node);
+}
+
+#endif /* CONFIG_ZBUS_RUNTIME_OBSERVERS_NODE_ALLOC_DYNAMIC */
```

Comment:
This entire `#else` block can now be removed?

Reply:
Yes, thanks. Done.